core/embed/bootloader/bootui.c

```diff
@@ -210,7 +210,7 @@ void ui_screen_boot(const vendor_header *const vhdr,
   uint32_t vimg_len = *(uint32_t *)(vhdr->vimg + 8);
 
   screen_boot(red_screen, vendor_str, vendor_str_len, hdr->version, vhdr->vimg,
-              vimg_len, wait);
+              vimg_len + 12, wait);
```

Comment:
It's quite unfortunate that we have to parse the TOIF header here. However, if it's necessary, it would be better to be explicit about what we're doing.

What about this:
```c
#define TOIF_LENGTH(ptr)  *(uint32_t *)((ptr) + 8) + 12;

uint32_t vimg_len = TOIF_LENGTH(vhdr->vimg)
screen_boot(red_screen, vendor_str, vendor_str_len, hdr->version, vhdr->vimg,
                 vimg_len, wait);
```c

Some checks are still missing here, but if the header integrity is checked beforehand, it might be okay.

Reply:
https://github.com/trezor/trezor-firmware/commit/5197e2b370d800ea81de572e92ea8cfaaafafb94